# Problem Set 1
**ECON 422: Macroeconomics & Machine Learning**

**Instructions.** Each problem is worth 1 point (0.1 point for basic credit; 0.3 points will be deducted for an incorrect
answer). For questions that ask for Python code, write code that can be run in a Jupyter notebook.

## Problem 1. Machine-learning framework

**(a)** Classify each task as **supervised**, **unsupervised**, or **reinforcement learning**:  
(i) predict house prices from property characteristics;  
(ii) group countries using macroeconomic indicators when there is no target variable;  
(iii) an agent learns which action to take from rewards received after interacting with an environment.

**(b)** In $y=g(X)$ and $\widehat y=h(X;\phi)$, briefly explain the roles of $g$, $h$, $\phi$, the candidate class $\mathcal H$, and the loss function.

**(c)** Why might economists use supervised machine learning even when prediction itself is not the final objective?

**Answer:**
i. Predicting housing prices from property characterstics is a textbook example of supervised learning. Supervised learning is defined as learning that contains a set of features, property characteristics, and a target, house prices, with the goal of learning relationships and predicting outcomes. Because predicting house prices with property characteristics involves using features to relate them to house prices with the goal of prediction, this is classified as supervised learning.

ii. Grouping countries using macroeconomic indicators without a target variable is unsupervised learning. Because grouping countries using macroeconomic indicators without a target contains no target variable, it cannot be supervised learning. It also adheres to the general purpose of unsupervised learning, which is to simply learn more about relationships between variables; in this case, it targets relationships between countries based off macroeconomic indicators.

iii. An agent learning which action to take based off rewards received from interacting with its environment is reinforcement learning. In reinforcement learning, an agent is put into an environment in which it makes certain choices and is rewarded based off desireable choices, thus steering the agent towards the desired choice and learning the desired rule that cumulates the most rewards. In this case, the agent learning via interactions with its environment and rewards based off those interactions fits the definition of reinforcement learning.

b. In these equations, g is the "ideal target function", that being the function of features that best models the relationship between the chosen features and the target variable; thus, this is the goal of the machine in that finding this function to model the target variable is the central goal of machine learning. The h is the approximation of g, in that this is what the machine will actually return in a similar way that an OLS model will return theta hat as an approximation for the actual relationship of the variable. Within this h function, the phi represents the model parameters that are estimated during the data analysis by the machine that represent the "theta" in the sense that they are the coefficients to the features within this approximation. The loss function represents the difference between the actual target function g and the approximated target function h, therefore computing the performance of the model at approximating the true relationships between the features and the target variable. H represents the collective of approximate functions in which one, h, is chosen as the best approximation of the target function, g. Thus, H contains many candidates of approximate functions acquired by different methods to be chosen as the approximation.

c. In economics, researchers are not only trying to forecast important economic variables but also trying to measure the causality of certain factors, often times the effectiveness of monetary and/or fiscal policies, or estimate difficult variables. So, economists might use supervised machine learning to conduct causal inference and estimate certain variables that might not be accurately explained by OLS. For estimating difficult variables, machine-learning methods can use the advantage of flexibility to estimate non-linear or convoluted relationships between features and target variables to then estimate the effects of those features on the target variable. This, then, can be used for causal inference by clarifying the treatment effect of a certain feature via the process of making other variables in the model easier to interpret and more accurate in effect. These advantages are why economists frequently use supervised machine learning methods for estimation and causal inference and not simply just prediction.

## Problem 2. Training and test samples with Python

A researcher has a cross section of households with features $X$ and outcome $y$.

**(a)** Explain the roles of the training sample and the test sample. Why should the test sample not be used to fit the model or choose its hyperparameters?

**Answer (a):**
a. Put simply, the main weakness of machine learning in general is that it tends to overfit the data that it uses to learn the relationships between variables. Thus, splitting the sample into training and testing samples is necessary to be able to both fit the data and test whether it can make accurate predictions based off data is has never seen before. In this, the training sample is the portion of the sample, often chosen randomly, that is used to directly train the model so the model can learn the relationships between the parameters and the target variable. The testing sample is the portion of the sample set aside in order to test the model that came from the training sample against data that it has never seen before to ensure that the model does not overfit itself onto the data it has seen and makes the most accurate predictions possible. Because of this role, the test sample should never be used to fit the model or choose hyperparameters; if the entire point of the test sample is to test the model against data it has never seen before, feeding it to the model to choose hyperparameters or fit the model defeats the purpose of the test sample by allowing the model to fit into the test sample in the process of learning and, therefore, makes the assessment of the model with the test sample less representative of the model's performance against unseen data.

**(b)** Write Python code using `train_test_split` to create a 75% training sample and a 25% test sample, with `random_state=42`. Assume arrays `X` and `y` already exist.

In [1]:
# Import necessary packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from IPython.display import display
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

np.random.seed(422)

pd.set_option("display.float_format", lambda x: f"{x:,.3f}")
# The train_test_split code that turns the assumed arrays X and y into training and testing samples with a 75% training to 25% testing distribution.
X_train, X_test, y_train, y_test = train_test_split(
    X, y, 
    test_size=0.25, 
    random_state=42)

# This code was partially written by Git Copilot that automatically filled code that I was going to write anyway based off the code used for this operation in our class handouts.

NameError: name 'X' is not defined

**(c)** For $y_{test}=(3,5,4,8)$ and $\widehat y=(2.5,4.5,5,7)$, compute the test MSE and RMSE. Also write Python code that performs the calculation.

In [1]:
# Import necessary packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from IPython.display import display
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
np.random.seed(422)
pd.set_option("display.float_format", lambda x: f"{x:,.3f}")

# Add the data specified in the question
y_test = [3, 5, 4, 8]
y_hat = [2.5, 4.5, 5, 7]

# Calculate the MSE and the RMSE
test_mse = mean_squared_error(y_test, y_hat)
test_rmse = np.sqrt(test_mse)

# Print it
print("Test MSE:", np.round(test_mse, 3))
print("Test RMSE:", np.round(test_rmse, 3))

# This code was partially written by Git Copilot that automatically filled code that I was going to write anyway based off the code used for this operation in our class handouts.

Test MSE: 0.625
Test RMSE: 0.791


**Numerical answer (c):**
The test MSE, as computed above, is 0.625 rounded to the nearest thousandth. The test RMSE is 0.791 rounded to the nearest thousandth.

## Problem 3. Hyperparameters and two-fold cross-validation

Consider the intercept-only model $y_i=\theta+\varepsilon_i$. Fold 1 outcomes are $(1,2,8)$ and Fold 2 outcomes are $(1,3,4)$. Let $a\in\{1,2\}$ in

$$
\widehat\theta(a)=\arg\min_{\theta}\sum_i |y_i-\theta|^a.
$$

For $a=1$, use the training-sample median; for $a=2$, use the training-sample mean. Use MSE as the common validation criterion.

**(a)** Compute the validation loss for $a=1$ and $a=2$ in each fold.

**(b)** Compute $\mathrm{CV}(1)$ and $\mathrm{CV}(2)$, and select $\widehat a$.

**(c)** Distinguish the model parameter $\theta$ from the hyperparameter $a$.

**Answer:**

a. The validation loss for a = 1 uses the training sample median to estimate theta hat while taking the absolute value of difference between y and theta, adding them together, and diving by the number of terms in the fold. Therefore, the loss for fold 1 when a = 1 and the median of the fold is 2 is 0.33 * [(1-2)^2 + (2-2)^2 + (8-2)^2] = 37/3. The loss of the second fold, a = 1 and median = 3, is 0.33 * [(1-3)^2 + (3-3)^2 + (4-3)^2] = 5/3. The loss when a = 2 uses the mean instead of the median, so the loss for fold 1 when a = 2 and the mean is (1+2+8)/3 = 11/3 is 0.33 * [(1 - 11/3)^2 + (2-11/3)^2 + (8-11/3)^2] = 86/9. The loss when a = 2 for fold 2, where the mean is (1+3+4)/3 = 8/3, is 0.33 * [(1-8/3)^2 + (3-8/3)^2 + (4-8/3)^2] = 23/9.

b. The CV(1) is the average of the losses for each fold given that the value of a = 1. Thus, CV(1) = 0.5 * (37/3 + 5/3) = 41/6, or approximately 6.83. The CV(2) is the average of the losses for each fold given that a = 2, so CV(2) = 0.5 * (23/9 + 86/9) = 109/18, or approximately 6.056. Because the CV when a = 2 is less than the CV when a = 1, then the selected value of a hat is 2.

c. The difference between the model parameter theta and the hyperparameter a lies in the function of the two. The model parameter theta is what is estimated during the fitting process to find the relationship between a variable and a target variable. On the other hand, the hyperparameter a determines how the fitting process goes on. In this case, theta is either the mean or median of the fold outcomes that is found during the fitting process and a is either 1 or 2 and controls how the fold outcomes estimated during the fitting process factor into the loss function, thus determining how the fitting process works.

## Problem 4. Regularization and tuning with Python

**(a)** State the penalty used by ridge, lasso, and elastic net. Which of these methods can set coefficients exactly equal to zero?

**(b)** What happens when $\lambda=0$, and what generally happens as $\lambda$ increases? Why are predictors generally standardized before regularized estimation?

**Answer (a)-(b):**
a. Ridge regression uses l_2 penalty, where lambda is multiplied by the sum of squared coefficients to add to the loss function as a penalty for larger coefficients, meaning that the coefficients get smaller as lambda gets larger to the point where the coefficients approach, but never touch, zero. Lasso uses an L1 penalty, where lambda is multiplied by the sum of the absolute value of the coefficients and added to the loss function to penalize larger coefficients; with this method, the coefficients can reach zero. The elastic net method combines the ridge and lasso methods by multiplying the lasso penalty by a factor alpha and the ridge penalty by a factor 1 - alpha, adding them both to the loss function as a penalty for large coefficients. Since the elastic net method combines the lasso and ridge methods, it can set coefficients to zero.

b. When lambda is zero, the loss function is basically OLS because the penalty of either of the methods becomes 0, thus being taken out of the equation. When lambda increases, the penalty for the coefficients increases and the coefficients infinitely approach zero for the ridge method, and reach zero for the elastic net and lasso methods. Because the penalties for the coefficients all have the same weight controlled by the hyperparameter lambda, the predictors must be standardized before regularization. That is, if the predictors are not on the same scale as each other, then the penalty will be weighed differently for different predictors, making them uncomparable and more difficult to interpret. Thus, the predictors should be standardized to the same scale before being regularized.

**(c)** Assume cross-sectional arrays `X` and `y` are available. Write Python code that standardizes the predictors and uses 5-fold cross-validation to choose ridge $\lambda$ from $\{0.01,0.1,1,10\}$ according to MSE.

In [ ]:
# Import other necessary packages
from sklearn.linear_model import (
    Ridge, Lasso, ElasticNet,
    RidgeCV, LassoCV, ElasticNetCV
)
from sklearn.metrics import mean_squared_error
from sklearn.model_selection import train_test_split, KFold
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
pd.set_option('display.float_format', lambda x: f'{x:0.4f}')

# Perform the ridge method and cross-validation
cv = KFold(n_splits=5, shuffle=True, random_state=42)
ridge_cv, ridge_summary = fit_and_measure(RidgeCV(alphas=np.logspace(-3, 3, 100), cv=cv))
ridge_summary.update(
    model = 'Ridge',
    selected_lambda = ridge_cv.named_steps['model'].alpha_,
    selected_l1_ratio = np.nan
)

## Problem 5. Bias-variance tradeoff

| Estimator | Irreducible noise $\sigma^2$ | Squared bias | Variance |
|---|---:|---:|---:|
| A | 2 | 0 | 9 |
| B | 2 | 1 | 4 |
| C | 2 | 5 | 1 |

**(a)** Compute MSE for A, B, and C using $\mathrm{MSE}=\sigma^2+\mathrm{Bias}^2+\mathrm{Variance}$.

**(b)** Which estimator has the smallest MSE?

**(c)** Interpret the comparison in terms of regularization and the bias-variance tradeoff.

**Answer:**

a. The MSE for A is found using MSE is equal to the irreducible noise plus the squared bias plus the variance. Thus, MSE(A) = 2 + 0 + 9 = 11. The MSE(B) uses the same formula, getting MSE(B) = 2 + 1 + 4 = 7. The MSE(C) also uses the same formula, giving MSE(C) = 2 + 5 + 1 = 8.

b. The estimator that has the smalleest MSE is the estimator B, with an MSE of 7 as opposed to estimator C's MSE of 8 and estimator A's MSE of 11.

c. In terms of regularization, regularization shrinks the estimators to a certain range of values, thus introducing some level of bias depending on the strength of the shrinkage but reducing the variance of the estimators. Thus, the MSE calculates the tradeoff between the reduced variance and the introduced bias, where both equally factor into the MSE and we are still trying to minimize the MSE based off the trade off. In this case, the MSE favors a medium-level shrinkage: estimator C shrinks the estimator too much, thus introducing more bias into the MSE while decreasing the variance relative to the estimator B by less than it increased the bias. Thus, estimator B gives a smaller MSE accounting for the bias-variance tradeoff. Relative to estimator A, which, given the squared bias value of 0, is simply OLS with no regularization, estimators B and C have smaller MSEs by undergoing the regularization process as the process decreases the variance of the estimators by a larger amount than they increase the squared bias.

## Problem 6. CART: choosing a split and coding it

A regression tree uses:

| $X$ | 2 | 3 | 4 | 7 | 8 | 9 |
|---|---:|---:|---:|---:|---:|---:|
| $Y$ | 4 | 4 | 3 | 0 | -1 | -2 |

At the root, consider cutoffs $s\in\{3.5,5.5,7.5\}$. Each child node predicts its mean outcome.

**(a)** Compute the root-node SSE and the post-split SSE for each cutoff. Which cutoff does CART select?

**(b)** Using only the selected split, what prediction is assigned to a new observation with $X=8.5$?

**Answer (a)-(b):**

a. The root-node SSE is goven from the sum of squared error where the y bar is the mean of the predictions and the and the y of i is the predicted y. The mean of Y is 1/6 * (4+4+3+0-1-2) = 4/3. Thus, the root-node SSE is [(4-4/3)^2 + (4-4/3)^2 + (3-4/3)^2 + (0-4/3)^2 + (-1-4/3)^2 + (-2-4/3)^2] = 106/3 which is approximately 35.33. The post-split SSE is the sum of SSEs after the split, where the SSE when X is less than the cutoff is added to the SSE when X is greater than the cutoff. So, the SSE when the cutoff is is X = 3.5 uses the means (4+4)/2 = 4 when less than 3.5 and (3+0-1-2)/4 = 0 when X > 3.5; therefore, the SSE for the cutoff X = 3.5 is [(4-4)^2 + (4-4)^2 + (3-0)^2 + (0-0)^2 + (-1-0)^2 + (-2-0)^2] = [9 + 1 + 4] = 14. For the cutoff X = 5.5, the mean when X < 5.5 is (4+4+3)/3 = 11/3 and the mean when X > 5.5 is (0-1-2)/3 = -1. So, the SSE when the cutoff is X = 5.5 is [(4-11/3)^2 + (4-11/3)^2 + (3-11/3)^2 + (0+1)^2 + (-1+1)^2 + (-2+1)^2] = 8/3 or about 2.67. The mean prior to the cutoff X = 7.5 is (4+4+3+0)/4 = 11/4 and the mean afterwards is (-1-2)/2 = -3/2. So, the SSE when the cutoff is X = 7.5 is [(4-11/4)^2 + (4-11/3)^2 + (3-11/4)^2 + (0-11/4)^2 + (-1+3/2)^2 + (-2+3/2)^2] = 45/4 or 11.25. Thus, with the lowest SSE of 8/3 or 2.67, CART will select the cutoff X = 5.5. 

b. The prediction assigned to X = 8.5 uses the mean of the values in the data when X > 5.5, the selected cutoff, since 8.5 > 5.5. Thus, Y(8.5) is predicted to be (0-1-2)/3 = -1.

**(c)** Write Python code that loops over the three candidate cutoffs, computes the post-split SSE, and returns the cutoff with the smallest SSE.

In [1]:
# Import Packages
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor, plot_tree
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.metrics import mean_squared_error
pd.set_option("display.float_format", lambda x: f"{x:,.3f}")

X = np.array([2, 3, 4, 7, 8, 9])
Y = np.array([4, 4, 3, 0, -1, -2])
cutoffs = [3.5, 5.5, 7.5]
cutoff_sse = []

# Create a loop that calculates the SSE for each cutoff value and returns the cutoff value that minimizes the SSE
for i in cutoffs:
    left_mask = X <= i
    right_mask = X > i
    left_mean = Y[left_mask].mean()
    right_mean = Y[right_mask].mean()
    sse = ((Y[left_mask] - left_mean) ** 2).sum() + ((Y[right_mask] - right_mean) ** 2).sum()
    cutoff_sse.append((i, sse))
    print(f"Cutoff: {i}, SSE: {np.round(sse, 3)}")

# Find the cutoff value that minimizes the SSE
best_cutoff = min(cutoff_sse, key=lambda x: x[1])[0]
print(f"Best cutoff: {best_cutoff} with SSE: {np.round(min(cutoff_sse, key=lambda x: x[1])[1], 3)}")

# This code was partially written by Git Copilot that automatically filled code to save time writing the loop conditions and SSE formula, along with the "min" function

Cutoff: 3.5, SSE: 14.0
Cutoff: 5.5, SSE: 2.667
Cutoff: 7.5, SSE: 11.25
Best cutoff: 5.5 with SSE: 2.667


## Problem 7. Random forest

**(a)** A tree uses bootstrap sample $(2,5,2,1,4)$ from observations $\{1,2,3,4,5\}$. Identify the out-of-bag observation. Then average the predictions $1.2,0.8,1.0,1.4$ from four trees.

**(b)** Why does a random forest use a random subset of predictors at each split?

**(c)** If $\sigma^2=9$, $\rho=0.20$, and $B=100$, compute

$$
\operatorname{Var}(\widehat f_{RF})\approx \rho\sigma^2+\frac{1-\rho}{B}\sigma^2.
$$

What component remains as $B\to\infty$?

**Answer:**

a. The out-of-bag observation for a bootstrap sample is simply the observation left over from the set of observations not used in the bootstrap sample. Given that the bootstrap sample is (2, 5, 2, 1, 4), the observation missing is 3; thus, the out-of-bag observation is 3. The average prediction from the four trees giving predictions 1.2, 0.8, 1.0, and 1.4 is (1.2 + 0.8 + 1.0 + 1.4) / 4 = 1.1.

b. When trees are not split randomly, there stands the risk of correlations between trees by nature of manual selection, similar to how manually selecting a dataset might introduce bias. Thus, splitting trees randomly helps reduce the correlation between trees, and thus reduces the variance of the predicted values predicted by the random forest.

c. If variance = 9, pairwise correlation is 0.2, and B terms = 100, then the Variance is 0.2 * 9 + (1 - 0.2) / 100 * 9 = 1.872. When B tends towards infinity, the denominator of the second term gets infinitely larger, thus making the entire second term infinitely smaller and tend towards zero. Thus, all that is left as B tends towards infinity is the term pairwise correlation times variance.

## Problem 8. Gradient boosting: one correction step

Suppose $X=(2,4,6,8)$ and $Y=(5,3,1,-1)$. Under squared-error loss, start from $\widehat f_0=\bar Y$. The first tree splits at $X<5$, and $\nu=0.5$.

**(a)** Compute $\widehat f_0$, the initial residuals, and the mean residual in each leaf.

**(b)** Compute the updated predictions $\widehat f_1(X)=\widehat f_0(X)+\nu h_1(X)$ and the new residuals.

**Answer (a)-(b):**

a. The mean of Y, Y bar, is f hat, so f hat = (5 + 3 + 1 - 1) / 4 = 2. The initial residuals use the formula Y of i minus f hat, giving r = (5 - 2, 3 - 2, 1 - 2, -1 - 2) = (3, 1, -1, -3). If the first tree splits at X < 5, then the mean residual when X < 5 is (3 + 1) / 2 = 4 / 2 = 2. The mean residual in the leaf where X > 5 is (-1 - (-3)) / 2 = (-1 - 3) / 2 = -4 / 2 = -2.

b. If nu = 0.5, then the updated predictions f hat of 1 = f hat of zero + 0.5 * (mean residuals at each leaf). Thus, the updated prediction when X < 5 is 2 + 0.5 * 2 = 2 + 1 = 3. The updated prediction when X > 5 is 2 - 0.5 * 2 = 1. The new residuals are (5 - 3, 3 - 3, 1 - 1, -1 - 1) = (2, 0, 0, -2) and the new mean residuals are (2 + 0 + 0 - 2)/ 4 = 0/ 4 = 0.

**(c)** Write NumPy code that reproduces parts (a)-(b). In one sentence, explain why boosting is described as learning the **remaining gap**.

In [5]:
# Set up the given data
X = np.array([2, 4, 6, 8])
Y = np.array([5,3, 1, -1])
nu = 0.5
split = 5

f_hat = np.mean(Y)
print("Initial f_hat:", f_hat)
residuals = Y - f_hat
print("Initial residuals:", residuals)
lower = X < split
upper = X > split
mean_residuals_lower = np.mean(residuals[lower])
print("Mean residuals X < 5:", mean_residuals_lower)
mean_residuals_upper = np.mean(residuals[upper])
print("Mean residuals X > 5:", mean_residuals_upper)
f_hat_lower_1 = nu * mean_residuals_lower + f_hat
f_hat_upper_1 = nu * mean_residuals_upper + f_hat
print("Updated f_hat for X < 5:", f_hat_lower_1)
print("Updated f_hat for X > 5:", f_hat_upper_1)
residuals_lower_1 = Y[lower] - f_hat_lower_1
residuals_upper_1 = Y[upper] - f_hat_upper_1
new_mean_residuals = np.mean([residuals_lower_1, residuals_upper_1])
print("New mean residuals:", new_mean_residuals)

# This code was partially filled by GitHub copilot that saved time on typing by automatically filling values I was already going to write.

# Gradient boosting is known as learning the remaining gap because it aims to update its predictions based off the gaps between the previous predictions and 
# the actual values, thus covering the gap by adding some value of the residuals to the prediction.

Initial f_hat: 2.0
Initial residuals: [ 3.  1. -1. -3.]
Mean residuals X < 5: 2.0
Mean residuals X > 5: -2.0
Updated f_hat for X < 5: 3.0
Updated f_hat for X > 5: 1.0
New mean residuals: 0.0


**Explanation (c):**

## Problem 9. Shapley values and SHAP

**(a)** What is SHAP and its implication?

For one observation,

$$
v_i(\varnothing)=1.5,\qquad v_i(X)=2.5,\qquad v_i(Z)=2.0,\qquad v_i(X,Z)=3.6.
$$

**(b)** Compute the SHAP contributions of $X$ and $Z$ by averaging each feature's marginal contribution across the two entry orders.

**(c)** Verify local accuracy, $\widehat f(x_i)=\phi_0+\phi_{iX}+\phi_{iZ}$, and interpret a positive SHAP value.

**Answer:**

## Problem 10. Cross-fitting and double/debiased machine learning

Consider the following estimation sample. The nuisance learners are regression trees that split at $X\leq 0.5$ and predict the training-sample mean within each leaf.

| $i$ | $X_i$ | $D_i$ | $Y_i$ | Fold |
|---:|---:|---:|---:|:---:|
| 1 | 0 | 1 | 6 | A |
| 2 | 0 | 1 | 6 | B |
| 3 | 0 | -1 | 2 | A |
| 4 | 0 | -1 | 2 | B |
| 5 | 1 | 11 | 20 | A |
| 6 | 1 | 11 | 20 | B |
| 7 | 1 | 9 | 16 | A |
| 8 | 1 | 9 | 16 | B |

Let $m_0(X)=E[D\mid X]$ and $\ell_0(X)=E[Y\mid X]$.

**(a)** Perform two-fold cross-fitting. Train on Fold B and calculate the out-of-fold nuisance predictions $\widehat m^{(-A)}(X_i)$ and $\widehat\ell^{(-A)}(X_i)$ for observations in Fold A. Then reverse the roles of the folds and report the cross-fitted values $\widehat m_i$ and $\widehat\ell_i$ for all eight observations.

**Answer (a):**

**(b)** Using these cross-fitted values, calculate $\widetilde D_i=D_i-\widehat m_i$ and $\widetilde Y_i=Y_i-\widehat\ell_i$, and then compute

$$
\widehat\theta=\frac{\sum_i\widetilde D_i\widetilde Y_i}{\sum_i\widetilde D_i^2}.
$$

Write Python code that reproduces your calculation.

**Numerical answer (b):**

**(c)** What problem does cross-fitting address? Why does DML additionally use Neyman orthogonality?

**Answer (c):**